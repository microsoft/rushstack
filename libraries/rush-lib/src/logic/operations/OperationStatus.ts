// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

/**
 * Enumeration defining potential states of an operation
 * @beta
 */
export enum OperationStatus {
  /**
   * The Operation is ready to execute. All its dependencies have succeeded.
   */
  Ready = 'READY',
  /**
   * The Operation is waiting for one or more dependencies to complete.
   */
  Waiting = 'WAITING',
  /**
   * The Operation is Queued
   */
  Queued = 'QUEUED',
  /**
   * The Operation is currently executing
   */
  Executing = 'EXECUTING',
  /**
   * The Operation completed successfully and did not write to standard output
   */
  Success = 'SUCCESS',
  /**
   * The Operation completed successfully, but wrote to standard output
   */
  SuccessWithWarning = 'SUCCESS WITH WARNINGS',
  /**
   * The Operation was skipped via the legacy incremental build logic
   */
  Skipped = 'SKIPPED',
  /**
   * The Operation had its outputs restored from the build cache
   */
  FromCache = 'FROM CACHE',
  /**
   * The Operation failed
   */
  Failure = 'FAILURE',
  /**
   * The Operation could not be executed because one or more of its dependencies failed
   */
  Blocked = 'BLOCKED',
  /**
   * The Operation was a no-op (for example, it had an empty script)
   */
  NoOp = 'NO OP'
}

/**
 * The set of statuses that are considered terminal.
 * @alpha
 */
export const TERMINAL_STATUSES: Set<OperationStatus> = new Set([
  OperationStatus.Success,
  OperationStatus.SuccessWithWarning,
  OperationStatus.Skipped,
  OperationStatus.Blocked,
  OperationStatus.FromCache,
  OperationStatus.Failure,
  OperationStatus.NoOp
]);
