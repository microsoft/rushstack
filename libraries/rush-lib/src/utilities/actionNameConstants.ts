// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

export const PURGE_ACTION_NAME: 'purge' = 'purge';
export const LINK_PACKAGE_ACTION_NAME: 'link-package' = 'link-package';
export const BRIDGE_PACKAGE_ACTION_NAME: 'bridge-package' = 'bridge-package';
