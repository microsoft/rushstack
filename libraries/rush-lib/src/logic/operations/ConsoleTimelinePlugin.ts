// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import type { ITerminal } from '@rushstack/terminal';
import { Colorize, PrintUtilities } from '@rushstack/terminal';

import type { IPhase } from '../../api/CommandLineConfiguration';
import type {
  ICreateOperationsContext,
  IPhasedCommandPlugin,
  PhasedCommandHooks
} from '../../pluginFramework/PhasedCommandHooks';
import type { IExecutionResult } from './IOperationExecutionResult';
import { OperationStatus } from './OperationStatus';
import type { CobuildConfiguration } from '../../api/CobuildConfiguration';

const PLUGIN_NAME: 'ConsoleTimelinePlugin' = 'ConsoleTimelinePlugin';

/* Sample output:
==============================================================================================================================
          @rushstack/tree-pattern (build) ###########-------------------------------------------------------------------- 3.3s
          @rushstack/eslint-patch (build) ########----------------------------------------------------------------------- 2.2s
           @rushstack/eslint-patch (test) -------%----------------------------------------------------------------------- 0.0s
@rushstack/eslint-plugin-security (build) ----------########################--------------------------------------------- 6.8s
@rushstack/eslint-plugin-packlets (build) ----------############################----------------------------------------- 8.1s
         @rushstack/eslint-plugin (build) ----------##############################--------------------------------------- 8.7s
           @rushstack/tree-pattern (test) ----------#####---------------------------------------------------------------- 1.2s
 @rushstack/eslint-plugin-security (test) ---------------------------------############---------------------------------- 3.3s
 @rushstack/eslint-plugin-packlets (test) -------------------------------------#####------------------------------------- 1.1s
         @rushstack/eslint-config (build) ---------------------------------------%--------------------------------------- 0.0s
          @rushstack/eslint-plugin (test) ---------------------------------------#############--------------------------- 3.8s
          @rushstack/eslint-config (test) ---------------------------------------%--------------------------------------- 0.0s
     @rushstack/node-core-library (build) ---------------------------------------################################-------- 9.5s
      @rushstack/node-core-library (test) ----------------------------------------------------------------------######### 2.2s
==============================================================================================================================
LEGEND:                                                                                                      Total Work: 50.3s
  [#] Success  [!] Failed/warnings  [%] Skipped/cached/no-op                                                 Wall Clock: 23.7s
                                                                                                       Max Parallelism Used: 5
                                                                                                     Avg Parallelism Used: 2.1
BY PHASE:
      _phase:build 38.6s
       _phase:test 11.7s
*/

/**
 * Phased command plugin that emits a timeline to the console.
 */
export class ConsoleTimelinePlugin implements IPhasedCommandPlugin {
  private readonly _terminal: ITerminal;

  public constructor(terminal: ITerminal) {
    this._terminal = terminal;
  }

  public apply(hooks: PhasedCommandHooks): void {
    hooks.afterExecuteOperations.tap(
      PLUGIN_NAME,
      (result: IExecutionResult, context: ICreateOperationsContext): void => {
        _printTimeline({
          terminal: this._terminal,
          result,
          cobuildConfiguration: context.cobuildConfiguration
        });
      }
    );
  }
}

/**
 * Timeline - a wider column width for printing the timeline summary
 */
const TIMELINE_WIDTH: number = 109;

/**
 * Timeline - symbols representing each operation status
 */
const TIMELINE_CHART_SYMBOLS: Record<OperationStatus, string> = {
  [OperationStatus.Waiting]: '?',
  [OperationStatus.Ready]: '?',
  [OperationStatus.Queued]: '?',
  [OperationStatus.Executing]: '?',
  [OperationStatus.Success]: '#',
  [OperationStatus.SuccessWithWarning]: '!',
  [OperationStatus.Failure]: '!',
  [OperationStatus.Blocked]: '.',
  [OperationStatus.Skipped]: '%',
  [OperationStatus.FromCache]: '%',
  [OperationStatus.NoOp]: '%'
};

const COBUILD_REPORTABLE_STATUSES: Set<OperationStatus> = new Set([
  OperationStatus.Success,
  OperationStatus.SuccessWithWarning,
  OperationStatus.Failure,
  OperationStatus.Blocked
]);

/**
 * Timeline - colorizer for each operation status
 */
const TIMELINE_CHART_COLORIZER: Record<OperationStatus, (string: string) => string> = {
  [OperationStatus.Waiting]: Colorize.yellow,
  [OperationStatus.Ready]: Colorize.yellow,
  [OperationStatus.Queued]: Colorize.yellow,
  [OperationStatus.Executing]: Colorize.yellow,
  [OperationStatus.Success]: Colorize.green,
  [OperationStatus.SuccessWithWarning]: Colorize.yellow,
  [OperationStatus.Failure]: Colorize.red,
  [OperationStatus.Blocked]: Colorize.red,
  [OperationStatus.Skipped]: Colorize.green,
  [OperationStatus.FromCache]: Colorize.green,
  [OperationStatus.NoOp]: Colorize.gray
};

interface ITimelineRecord {
  startTime: number;
  endTime: number;
  durationString: string;
  name: string;
  status: OperationStatus;
  isExecuteByOtherCobuildRunner: boolean;
}

/**
 * @internal
 */
export interface IPrintTimelineParameters {
  terminal: ITerminal;
  result: IExecutionResult;
  cobuildConfiguration: CobuildConfiguration | undefined;
}

/**
 * Print a more detailed timeline and analysis of CPU usage for the build.
 * @internal
 */
export function _printTimeline({ terminal, result, cobuildConfiguration }: IPrintTimelineParameters): void {
  //
  // Gather the operation records we'll be displaying. Do some inline max()
  // finding to reduce the number of times we need to loop through operations.
  //

  const durationByPhase: Map<IPhase, number> = new Map();

  const data: ITimelineRecord[] = [];
  let longestNameLength: number = 0;
  let longestDurationLength: number = 0;
  let allStart: number = Infinity;
  let allEnd: number = -Infinity;
  let workDuration: number = 0;

  for (const [operation, operationResult] of result.operationResults) {
    if (operationResult.silent) {
      continue;
    }

    const { stopwatch } = operationResult;

    const { startTime, endTime } = stopwatch;

    if (startTime && endTime) {
      const nameLength: number = operation.name?.length || 0;
      if (nameLength > longestNameLength) {
        longestNameLength = nameLength;
      }

      const { duration } = stopwatch;
      const durationString: string = duration.toFixed(1);
      const durationLength: number = durationString.length;
      if (durationLength > longestDurationLength) {
        longestDurationLength = durationLength;
      }

      if (endTime > allEnd) {
        allEnd = endTime;
      }
      if (startTime < allStart) {
        allStart = startTime;
      }
      workDuration += duration;

      const { associatedPhase } = operation;

      if (associatedPhase) {
        durationByPhase.set(associatedPhase, (durationByPhase.get(associatedPhase) || 0) + duration);
      }

      data.push({
        startTime,
        endTime,
        durationString,
        name: operation.name!,
        status: operationResult.status,
        isExecuteByOtherCobuildRunner:
          !!operationResult.cobuildRunnerId &&
          operationResult.cobuildRunnerId !== cobuildConfiguration?.cobuildRunnerId
      });
    }
  }

  data.sort((a, b) => a.startTime - b.startTime);

  //
  // Determine timing for all tasks (wall clock and execution times)
  //

  const allDuration: number = allEnd - allStart;
  const allDurationSeconds: number = allDuration / 1000;

  //
  // Do some calculations to determine what size timeline chart we need.
  //

  const maxWidth: number = PrintUtilities.getConsoleWidth() || TIMELINE_WIDTH;
  const chartWidth: number = maxWidth - longestNameLength - longestDurationLength - 4;
  //
  // Loop through all operations, assembling some statistics about operations and
  // phases, if applicable.
  //

  const busyCpus: number[] = [];
  function getOpenCPU(time: number): number {
    const len: number = busyCpus.length;
    for (let i: number = 0; i < len; i++) {
      if (busyCpus[i] <= time) {
        return i;
      }
    }
    return len;
  }

  // Start with a newline
  terminal.writeLine('');
  terminal.writeLine('='.repeat(maxWidth));

  let hasCobuildSymbol: boolean = false;

  function getChartSymbol(record: ITimelineRecord): string {
    const { isExecuteByOtherCobuildRunner, status } = record;
    if (isExecuteByOtherCobuildRunner && COBUILD_REPORTABLE_STATUSES.has(status)) {
      hasCobuildSymbol = true;
      return 'C';
    }
    return TIMELINE_CHART_SYMBOLS[status];
  }

  for (const record of data) {
    const { startTime, endTime, durationString, name, status } = record;
    // Track busy CPUs
    const openCpu: number = getOpenCPU(startTime);
    busyCpus[openCpu] = endTime;

    // Build timeline chart
    const startIdx: number = Math.floor(((startTime - allStart) * chartWidth) / allDuration);
    const endIdx: number = Math.floor(((endTime - allStart) * chartWidth) / allDuration);
    const length: number = endIdx - startIdx + 1;

    const chart: string =
      Colorize.gray('-'.repeat(startIdx)) +
      TIMELINE_CHART_COLORIZER[status](getChartSymbol(record).repeat(length)) +
      Colorize.gray('-'.repeat(chartWidth - endIdx));
    terminal.writeLine(
      `${Colorize.cyan(name.padStart(longestNameLength))} ${chart} ${Colorize.white(
        durationString.padStart(longestDurationLength) + 's'
      )}`
    );
  }

  terminal.writeLine('='.repeat(maxWidth));

  //
  // Format legend and summary areas
  //

  const usedCpus: number = busyCpus.length;

  const legend: string[] = [
    'LEGEND:',
    '  [#] Success  [!] Failed/warnings  [%] Skipped/cached/no-op',
    '',
    ''
  ];
  if (hasCobuildSymbol) {
    legend[2] = '  [C] Cobuild';
  }

  const summary: string[] = [
    `Total Work: ${workDuration.toFixed(1)}s`,
    `Wall Clock: ${allDurationSeconds.toFixed(1)}s`,
    `Max Parallelism Used: ${usedCpus}`,
    `Avg Parallelism Used: ${(workDuration / allDurationSeconds).toFixed(1)}`
  ];

  terminal.writeLine(legend[0] + summary[0].padStart(maxWidth - legend[0].length));
  terminal.writeLine(legend[1] + summary[1].padStart(maxWidth - legend[1].length));
  terminal.writeLine(legend[2] + summary[2].padStart(maxWidth - legend[2].length));
  terminal.writeLine(legend[3] + summary[3].padStart(maxWidth - legend[3].length));

  //
  // Include time-by-phase, if phases are enabled
  //

  if (durationByPhase.size > 0) {
    terminal.writeLine('BY PHASE:');

    let maxPhaseName: number = 16;
    for (const phase of durationByPhase.keys()) {
      const len: number = phase.name.length;
      if (len > maxPhaseName) {
        maxPhaseName = len;
      }
    }

    for (const [phase, duration] of durationByPhase.entries()) {
      terminal.writeLine(`  ${Colorize.cyan(phase.name.padStart(maxPhaseName))} ${duration.toFixed(1)}s`);
    }
  }

  terminal.writeLine('');
}
