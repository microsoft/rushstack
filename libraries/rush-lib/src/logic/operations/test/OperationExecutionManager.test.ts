// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

// The TaskExecutionManager prints "x.xx seconds" in TestRunner.test.ts.snap; ensure that the Stopwatch timing is deterministic
jest.mock('../../../utilities/Utilities');

jest.mock('@rushstack/terminal', () => {
  const originalModule = jest.requireActual('@rushstack/terminal');
  return {
    ...originalModule,
    ConsoleTerminalProvider: {
      ...originalModule.ConsoleTerminalProvider,
      supportsColor: true
    }
  };
});

import { Terminal } from '@rushstack/terminal';
import { CollatedTerminal } from '@rushstack/stream-collator';
import { MockWritable, PrintUtilities } from '@rushstack/terminal';

import {
  OperationExecutionManager,
  type IOperationExecutionManagerOptions
} from '../OperationExecutionManager';
import { _printOperationStatus } from '../OperationResultSummarizerPlugin';
import { _printTimeline } from '../ConsoleTimelinePlugin';
import { OperationStatus } from '../OperationStatus';
import { Operation } from '../Operation';
import { Utilities } from '../../../utilities/Utilities';
import type { IOperationRunner } from '../IOperationRunner';
import { MockOperationRunner } from './MockOperationRunner';
import type { IExecutionResult, IOperationExecutionResult } from '../IOperationExecutionResult';
import { CollatedTerminalProvider } from '../../../utilities/CollatedTerminalProvider';

const mockGetTimeInMs: jest.Mock = jest.fn();
Utilities.getTimeInMs = mockGetTimeInMs;

let mockTimeInMs: number = 0;
mockGetTimeInMs.mockImplementation(() => {
  mockTimeInMs += 100;
  return mockTimeInMs;
});

const mockWritable: MockWritable = new MockWritable();
const mockTerminal: Terminal = new Terminal(new CollatedTerminalProvider(new CollatedTerminal(mockWritable)));

function createExecutionManager(
  executionManagerOptions: IOperationExecutionManagerOptions,
  operationRunner: IOperationRunner
): OperationExecutionManager {
  const operation: Operation = new Operation({
    runner: operationRunner,
    logFilenameIdentifier: 'operation'
  });

  return new OperationExecutionManager(new Set([operation]), executionManagerOptions);
}

describe(OperationExecutionManager.name, () => {
  let executionManager: OperationExecutionManager;
  let executionManagerOptions: IOperationExecutionManagerOptions;

  beforeEach(() => {
    jest.spyOn(PrintUtilities, 'getConsoleWidth').mockReturnValue(90);
    mockWritable.reset();
  });

  describe('Error logging', () => {
    beforeEach(() => {
      executionManagerOptions = {
        quietMode: false,
        debugMode: false,
        parallelism: 1,
        changedProjectsOnly: false,
        destination: mockWritable
      };
    });

    it('printedStderrAfterError', async () => {
      executionManager = createExecutionManager(
        executionManagerOptions,
        new MockOperationRunner('stdout+stderr', async (terminal: CollatedTerminal) => {
          terminal.writeStdoutLine('Build step 1\n');
          terminal.writeStderrLine('Error: step 1 failed\n');
          return OperationStatus.Failure;
        })
      );

      const result: IExecutionResult = await executionManager.executeAsync();
      _printOperationStatus(mockTerminal, false, result);
      expect(result.status).toEqual(OperationStatus.Failure);
      expect(result.operationResults.size).toEqual(1);
      const firstResult: IOperationExecutionResult = result.operationResults.values().next().value;
      expect(firstResult.status).toEqual(OperationStatus.Failure);

      const allMessages: string = mockWritable.getAllOutput();
      expect(allMessages).toContain('Error: step 1 failed');
      expect(mockWritable.getFormattedChunks()).toMatchSnapshot();
    });

    it('printedStdoutAfterErrorWithEmptyStderr', async () => {
      executionManager = createExecutionManager(
        executionManagerOptions,
        new MockOperationRunner('stdout only', async (terminal: CollatedTerminal) => {
          terminal.writeStdoutLine('Build step 1\n');
          terminal.writeStdoutLine('Error: step 1 failed\n');
          return OperationStatus.Failure;
        })
      );

      const result: IExecutionResult = await executionManager.executeAsync();
      _printOperationStatus(mockTerminal, false, result);
      expect(result.status).toEqual(OperationStatus.Failure);
      expect(result.operationResults.size).toEqual(1);
      const firstResult: IOperationExecutionResult = result.operationResults.values().next().value;
      expect(firstResult.status).toEqual(OperationStatus.Failure);

      const allOutput: string = mockWritable.getAllOutput();
      expect(allOutput).toMatch(/Build step 1/);
      expect(allOutput).toMatch(/Error: step 1 failed/);
      expect(mockWritable.getFormattedChunks()).toMatchSnapshot();
    });
  });

  describe('Blocking', () => {
    it('Failed operations block', async () => {
      const failingOperation = new Operation({
        runner: new MockOperationRunner('fail', async () => {
          return OperationStatus.Failure;
        }),
        logFilenameIdentifier: 'fail'
      });

      const blockedRunFn: jest.Mock = jest.fn();

      const blockedOperation = new Operation({
        runner: new MockOperationRunner('blocked', blockedRunFn),
        logFilenameIdentifier: 'blocked'
      });

      blockedOperation.addDependency(failingOperation);

      const manager: OperationExecutionManager = new OperationExecutionManager(
        new Set([failingOperation, blockedOperation]),
        {
          quietMode: false,
          debugMode: false,
          parallelism: 1,
          changedProjectsOnly: false,
          destination: mockWritable
        }
      );

      const result = await manager.executeAsync();
      expect(result.status).toEqual(OperationStatus.Failure);
      expect(blockedRunFn).not.toHaveBeenCalled();
      expect(result.operationResults.size).toEqual(2);
      expect(result.operationResults.get(failingOperation)?.status).toEqual(OperationStatus.Failure);
      expect(result.operationResults.get(blockedOperation)?.status).toEqual(OperationStatus.Blocked);
    });
  });

  describe('Warning logging', () => {
    describe('Fail on warning', () => {
      beforeEach(() => {
        executionManagerOptions = {
          quietMode: false,
          debugMode: false,
          parallelism: 1,
          changedProjectsOnly: false,
          destination: mockWritable
        };
      });

      it('Logs warnings correctly', async () => {
        executionManager = createExecutionManager(
          executionManagerOptions,
          new MockOperationRunner('success with warnings (failure)', async (terminal: CollatedTerminal) => {
            terminal.writeStdoutLine('Build step 1\n');
            terminal.writeStdoutLine('Warning: step 1 succeeded with warnings\n');
            return OperationStatus.SuccessWithWarning;
          })
        );

        const result: IExecutionResult = await executionManager.executeAsync();
        _printOperationStatus(mockTerminal, false, result);
        expect(result.status).toEqual(OperationStatus.SuccessWithWarning);
        expect(result.operationResults.size).toEqual(1);
        const firstResult: IOperationExecutionResult = result.operationResults.values().next().value;
        expect(firstResult.status).toEqual(OperationStatus.SuccessWithWarning);

        const allMessages: string = mockWritable.getAllOutput();
        expect(allMessages).toContain('Build step 1');
        expect(allMessages).toContain('step 1 succeeded with warnings');
        expect(mockWritable.getFormattedChunks()).toMatchSnapshot();
      });
    });

    describe('Success on warning', () => {
      beforeEach(() => {
        executionManagerOptions = {
          quietMode: false,
          debugMode: false,
          parallelism: 1,
          changedProjectsOnly: false,
          destination: mockWritable
        };
      });

      it('Logs warnings correctly', async () => {
        executionManager = createExecutionManager(
          executionManagerOptions,
          new MockOperationRunner(
            'success with warnings (success)',
            async (terminal: CollatedTerminal) => {
              terminal.writeStdoutLine('Build step 1\n');
              terminal.writeStdoutLine('Warning: step 1 succeeded with warnings\n');
              return OperationStatus.SuccessWithWarning;
            },
            /* warningsAreAllowed */ true
          )
        );

        const result: IExecutionResult = await executionManager.executeAsync();
        _printOperationStatus(mockTerminal, false, result);
        expect(result.status).toEqual(OperationStatus.Success);
        expect(result.operationResults.size).toEqual(1);
        const firstResult: IOperationExecutionResult = result.operationResults.values().next().value;
        expect(firstResult.status).toEqual(OperationStatus.SuccessWithWarning);
        const allMessages: string = mockWritable.getAllOutput();
        expect(allMessages).toContain('Build step 1');
        expect(allMessages).toContain('Warning: step 1 succeeded with warnings');
        expect(mockWritable.getFormattedChunks()).toMatchSnapshot();
      });

      it('logs warnings correctly with --timeline option', async () => {
        executionManager = createExecutionManager(
          executionManagerOptions,
          new MockOperationRunner(
            'success with warnings (success)',
            async (terminal: CollatedTerminal) => {
              terminal.writeStdoutLine('Build step 1\n');
              terminal.writeStdoutLine('Warning: step 1 succeeded with warnings\n');
              return OperationStatus.SuccessWithWarning;
            },
            /* warningsAreAllowed */ true
          )
        );

        const result: IExecutionResult = await executionManager.executeAsync();
        _printTimeline({ terminal: mockTerminal, result, cobuildConfiguration: undefined });
        _printOperationStatus(mockTerminal, false, result);
        const allMessages: string = mockWritable.getAllOutput();
        expect(allMessages).toContain('Build step 1');
        expect(allMessages).toContain('Warning: step 1 succeeded with warnings');
        expect(mockWritable.getFormattedChunks()).toMatchSnapshot();
      });
    });
  });
});
