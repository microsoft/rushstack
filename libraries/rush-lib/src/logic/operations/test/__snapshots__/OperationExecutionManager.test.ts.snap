// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OperationExecutionManager Error logging printedStderrAfterError 1`] = `
Array [
  Object {
    "kind": "O",
    "text": "Selected 1 operation:
",
  },
  Object {
    "kind": "O",
    "text": "  stdout+stderr
",
  },
  Object {
    "kind": "O",
    "text": "
",
  },
  Object {
    "kind": "O",
    "text": "Executing a maximum of 1 simultaneous processes...
",
  },
  Object {
    "kind": "O",
    "text": "
[gray]==[[default] [cyan]stdout+stderr[default] [gray]]================================================[[default] [white]1 of 1[default] [gray]]==[default]
",
  },
  Object {
    "kind": "O",
    "text": "
",
  },
  Object {
    "kind": "O",
    "text": "Build step 1

",
  },
  Object {
    "kind": "E",
    "text": "Error: step 1 failed

",
  },
  Object {
    "kind": "E",
    "text": "[red]\\"stdout+stderr\\" failed to build.[default]
",
  },
  Object {
    "kind": "O",
    "text": "


",
  },
  Object {
    "kind": "O",
    "text": "[gray]==[[default] [red]FAILURE: 1 operation[default] [gray]]=====================================================[default]

",
  },
  Object {
    "kind": "O",
    "text": "[gray]--[[default] [red]FAILURE: stdout+stderr[default] [gray]]---------------------------------[[default] [white]0.10 seconds[default] [gray]]--[default]

",
  },
  Object {
    "kind": "O",
    "text": "
",
  },
  Object {
    "kind": "O",
    "text": "
",
  },
  Object {
    "kind": "E",
    "text": "[red]Operations failed.
[default]
",
  },
]
`;

exports[`OperationExecutionManager Error logging printedStdoutAfterErrorWithEmptyStderr 1`] = `
Array [
  Object {
    "kind": "O",
    "text": "Selected 1 operation:
",
  },
  Object {
    "kind": "O",
    "text": "  stdout only
",
  },
  Object {
    "kind": "O",
    "text": "
",
  },
  Object {
    "kind": "O",
    "text": "Executing a maximum of 1 simultaneous processes...
",
  },
  Object {
    "kind": "O",
    "text": "
[gray]==[[default] [cyan]stdout only[default] [gray]]==================================================[[default] [white]1 of 1[default] [gray]]==[default]
",
  },
  Object {
    "kind": "O",
    "text": "
",
  },
  Object {
    "kind": "O",
    "text": "Build step 1

",
  },
  Object {
    "kind": "O",
    "text": "Error: step 1 failed

",
  },
  Object {
    "kind": "E",
    "text": "[red]\\"stdout only\\" failed to build.[default]
",
  },
  Object {
    "kind": "O",
    "text": "


",
  },
  Object {
    "kind": "O",
    "text": "[gray]==[[default] [red]FAILURE: 1 operation[default] [gray]]=====================================================[default]

",
  },
  Object {
    "kind": "O",
    "text": "[gray]--[[default] [red]FAILURE: stdout only[default] [gray]]-----------------------------------[[default] [white]0.10 seconds[default] [gray]]--[default]

",
  },
  Object {
    "kind": "O",
    "text": "
",
  },
  Object {
    "kind": "O",
    "text": "
",
  },
  Object {
    "kind": "E",
    "text": "[red]Operations failed.
[default]
",
  },
]
`;

exports[`OperationExecutionManager Warning logging Fail on warning Logs warnings correctly 1`] = `
Array [
  Object {
    "kind": "O",
    "text": "Selected 1 operation:
",
  },
  Object {
    "kind": "O",
    "text": "  success with warnings (failure)
",
  },
  Object {
    "kind": "O",
    "text": "
",
  },
  Object {
    "kind": "O",
    "text": "Executing a maximum of 1 simultaneous processes...
",
  },
  Object {
    "kind": "O",
    "text": "
[gray]==[[default] [cyan]success with warnings (failure)[default] [gray]]==============================[[default] [white]1 of 1[default] [gray]]==[default]
",
  },
  Object {
    "kind": "O",
    "text": "
",
  },
  Object {
    "kind": "O",
    "text": "Build step 1

",
  },
  Object {
    "kind": "O",
    "text": "Warning: step 1 succeeded with warnings

",
  },
  Object {
    "kind": "E",
    "text": "[yellow]\\"success with warnings (failure)\\" completed with warnings in 0.10 seconds.[default]
",
  },
  Object {
    "kind": "O",
    "text": "


",
  },
  Object {
    "kind": "O",
    "text": "[gray]==[[default] [yellow]SUCCESS WITH WARNINGS: 1 operation[default] [gray]]=======================================[default]

",
  },
  Object {
    "kind": "O",
    "text": "[gray]--[[default] [yellow]WARNING: success with warnings (failure)[default] [gray]]---------------[[default] [white]0.20 seconds[default] [gray]]--[default]

",
  },
  Object {
    "kind": "O",
    "text": "
",
  },
  Object {
    "kind": "O",
    "text": "
",
  },
  Object {
    "kind": "E",
    "text": "[yellow]Operations succeeded with warnings.
[default]
",
  },
]
`;

exports[`OperationExecutionManager Warning logging Success on warning Logs warnings correctly 1`] = `
Array [
  Object {
    "kind": "O",
    "text": "Selected 1 operation:
",
  },
  Object {
    "kind": "O",
    "text": "  success with warnings (success)
",
  },
  Object {
    "kind": "O",
    "text": "
",
  },
  Object {
    "kind": "O",
    "text": "Executing a maximum of 1 simultaneous processes...
",
  },
  Object {
    "kind": "O",
    "text": "
[gray]==[[default] [cyan]success with warnings (success)[default] [gray]]==============================[[default] [white]1 of 1[default] [gray]]==[default]
",
  },
  Object {
    "kind": "O",
    "text": "
",
  },
  Object {
    "kind": "O",
    "text": "Build step 1

",
  },
  Object {
    "kind": "O",
    "text": "Warning: step 1 succeeded with warnings

",
  },
  Object {
    "kind": "E",
    "text": "[yellow]\\"success with warnings (success)\\" completed with warnings in 0.10 seconds.[default]
",
  },
  Object {
    "kind": "O",
    "text": "


",
  },
  Object {
    "kind": "O",
    "text": "[gray]==[[default] [yellow]SUCCESS WITH WARNINGS: 1 operation[default] [gray]]=======================================[default]

",
  },
  Object {
    "kind": "O",
    "text": "[gray]--[[default] [yellow]WARNING: success with warnings (success)[default] [gray]]---------------[[default] [white]0.20 seconds[default] [gray]]--[default]

",
  },
  Object {
    "kind": "O",
    "text": "
",
  },
  Object {
    "kind": "O",
    "text": "
",
  },
]
`;

exports[`OperationExecutionManager Warning logging Success on warning logs warnings correctly with --timeline option 1`] = `
Array [
  Object {
    "kind": "O",
    "text": "Selected 1 operation:
",
  },
  Object {
    "kind": "O",
    "text": "  success with warnings (success)
",
  },
  Object {
    "kind": "O",
    "text": "
",
  },
  Object {
    "kind": "O",
    "text": "Executing a maximum of 1 simultaneous processes...
",
  },
  Object {
    "kind": "O",
    "text": "
[gray]==[[default] [cyan]success with warnings (success)[default] [gray]]==============================[[default] [white]1 of 1[default] [gray]]==[default]
",
  },
  Object {
    "kind": "O",
    "text": "
",
  },
  Object {
    "kind": "O",
    "text": "Build step 1

",
  },
  Object {
    "kind": "O",
    "text": "Warning: step 1 succeeded with warnings

",
  },
  Object {
    "kind": "E",
    "text": "[yellow]\\"success with warnings (success)\\" completed with warnings in 0.10 seconds.[default]
",
  },
  Object {
    "kind": "O",
    "text": "
",
  },
  Object {
    "kind": "O",
    "text": "==========================================================================================
",
  },
  Object {
    "kind": "O",
    "text": "[cyan]success with warnings (success)[default] [gray][default][yellow]!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!![default][gray][default] [white]0.2s[default]
",
  },
  Object {
    "kind": "O",
    "text": "==========================================================================================
",
  },
  Object {
    "kind": "O",
    "text": "LEGEND:                                                                   Total Work: 0.2s
",
  },
  Object {
    "kind": "O",
    "text": "  [#] Success  [!] Failed/warnings  [%] Skipped/cached/no-op              Wall Clock: 0.2s
",
  },
  Object {
    "kind": "O",
    "text": "                                                                   Max Parallelism Used: 1
",
  },
  Object {
    "kind": "O",
    "text": "                                                                 Avg Parallelism Used: 1.0
",
  },
  Object {
    "kind": "O",
    "text": "BY PHASE:
",
  },
  Object {
    "kind": "O",
    "text": "  [cyan]           phase[default] 0.2s
",
  },
  Object {
    "kind": "O",
    "text": "
",
  },
  Object {
    "kind": "O",
    "text": "


",
  },
  Object {
    "kind": "O",
    "text": "[gray]==[[default] [yellow]SUCCESS WITH WARNINGS: 1 operation[default] [gray]]=======================================[default]

",
  },
  Object {
    "kind": "O",
    "text": "[gray]--[[default] [yellow]WARNING: success with warnings (success)[default] [gray]]---------------[[default] [white]0.20 seconds[default] [gray]]--[default]

",
  },
  Object {
    "kind": "O",
    "text": "
",
  },
  Object {
    "kind": "O",
    "text": "
",
  },
]
`;
